_posts/2019-08-26-how-to-use-gruntwork-infrastructure-as-code-library.adoc

```diff
@@ -1399,6 +1409,272 @@ module "vpc" {
+
+In the future, Gruntwork may offer modules through  a private Terraform registry to further simplify the process.
+
+[[terragrunt_tfc_tfe]]
+==== Using Terragrunt with TFC and TFE
+Configuring Terragrunt to work with TFC/TFE is a bit more involved. As mentioned previously, TFC and TFE execute
+`terraform` commands under the hood, and the executors that run these commands aren't able to execute the Terragrunt
+binary. Running Terragrunt directly through the TFC UI isn't an option.
+
+However, Terragrunt can rely on TFC's https://www.terraform.io/docs/cloud/run/cli.html#implicit-workspace-creation[implicit workspace creation] feature to automatically create workspaces when Terragrunt runs. This feature allows you to use Terragrunt to invoke TFC runs from your local CLI. Once you're up and running, the state will be maintained in TFC like any other workspace, and the standard plan and apply workflow will function normally. There are several steps involved to make this work:
+
+1. Generate a `backend` block to use TFC/TFE for remote state storage and execution.
+1. Generate a `terraform.auto.tfvars` for submitting variables.
+1. Configure authentication to providers.
+
+In the sections below we will cover each step in turn.
+
+===== Setting up
+
+Consider the following Terragrunt directory structure:
+
+[source]
+----
+.
+├── dev
+│   ├── account.hcl
+│   └── us-east-1
+│       ├── region.hcl
+│       ├── sns
+│       │   └── terragrunt.hcl
+│       └── sqs
+│           └── terragrunt.hcl
+├── common.hcl
+└── terragrunt.hcl
+----
+
+For simplicity, the example shows a single environment, `dev`, with two modules: `sqs` and `sns`. The pattern could be expanded to more environments and regions by adding to the hierarchy as needed.
+
+===== Configuring the backend
+
+For Terragrunt to use TFC as a remote executor, you'll need a `backend` block. The backend is what determines where Terraform reads state and how execution is handled. In this case, you'll configure TFC as a remote in the backend block. The Terragrunt https://terragrunt.gruntwork.io/docs/reference/config-blocks-and-attributes/#generate:[`generate` block] can generate the block dynamically so you don't need to include one in each module.
+
+Terraform https://www.terraform.io/docs/state/workspaces.html[workspaces] store the state associated with a given backend. In the Terragrunt configuration hierarchy defined above, each module (`sns` and `sqs`) is a separate workspace and must have its own name in the `backend` block. There are several ways to do this with Terragrunt. The snippet below shows one approach: use the account name (defined in `account.hcl`) as a prefix, and use the directory name of the module as the suffix.
+
+The following snippet is placed in `terragrunt.hcl` at the root of the hierarchy.
+
+[source,hcl]
+----
+locals {
+  tfc_hostname     = "app.terraform.io"
```

Comment:
NIT: put a comment here mentioning to change this to their custom URL if using TFE.